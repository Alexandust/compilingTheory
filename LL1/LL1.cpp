/*************************************************************************
	> File Name: LL1.cpp
	  > Author: Netcan
	  > Blog: http://www.netcan666.com
	  > Mail: 1469709759@qq.com
	  > Created Time: 2016-10-09 日 19:23:18 CST

                                                    __----~~~~~~~~~~~------___
                                   .  .   ~~//====......          __--~ ~~
                   -.            \_|//     |||\\  ~~~~~~::::... /~
                ___-==_       _-~o~  \/    |||  \\            _/~~-
        __---~~~.==~||\=_    -_--~/_-~|-   |\\   \\        _/~
    _-~~     .=~    |  \\-_    '-~7  /-   /  ||    \      /
  .~       .~       |   \\ -_    /  /-   /   ||      \   /
 /  ____  /         |     \\ ~-_/  /|- _/   .||       \ /
 |~~    ~~|--~~~~--_ \     ~==-/   | \~--===~~        .\
          '         ~-|      /|    |-~\~~       __--~~
                      |-~~-_/ |    |   ~\_   _-~            /\
                           /  \     \__   \/~                \__
                       _--~ _/ | .-~~____--~-/                  ~~==.
                      ((->/~   '.|||' -_|    ~~-/ ,              . _||
                                 -_     ~\      ~~---l__i__i__i--~~_/
                                 _-~-__   ~)  \--______________--~~
                               //.-~~~-~_--~- |-------~~~~~~~~
                                      //.-~~~--\
						神兽保佑，代码无BUG!
*************************************************************************/

#include "LL1.h"

Prod::Prod(const string &in) {
	prod = in;
	isValid = false;
	split();
}

bool Prod::split() {
	if(prod.length() < 4) return false;
	if(cut(1, 3) == "->" && isupper(prod[0])) // A->...则noTerminal = A
		noTerminal = prod[0];
	else return false;
	for(unsigned int i=0; i<prod.length(); ++i) { // 提取终结符、非终结符
		char c = prod[i];
		if(isupper(c)) Vn.insert(c);
		else if(c!='|' && !(c=='-' && prod[i+1] == '>' && ++i)) Vt.insert(c);
	}

	for(unsigned int i=3; i<prod.length(); ++i) { // 提取候选式
		unsigned int j;
		for(j=i+1; j<prod.length() && prod[j] != '|'; ++j);
		selection.insert(cut(i, j));
		i = j;
	}

	return isValid = true;
}

bool LL1::addProd(const Prod &prod) {
	if(prod.isValid) {
		vector<Prod>::iterator it = find(G.begin(), G.end(), prod.noTerminal);
		if(it != G.end()) {   // 找到产生式
			it->selection.insert(prod.selection.begin(), prod.selection.end()); // 追加候选式
			for(auto s: prod.selection)
				it->prod += ("|" + s); // 追加候选式
		}
		else  // 没找到
			G.push_back(prod);
		VN.insert(prod.Vn.begin(), prod.Vn.end());
		VT.insert(prod.Vt.begin(), prod.Vt.end());
		return true;
	}
	else return false;
}

set<char> LL1::first(const string &s) {
	Prod prod = Prod(s);
	if(prod.isValid) { // 产生式
		if(FIRST[prod.noTerminal].size() != 0) return FIRST[prod.noTerminal];
		for(auto sel:prod.selection) { // 候选式
			set<char> f = first(sel); // 求出候选式的first集合
			FIRST[prod.noTerminal].insert(f.begin(), f.end());
		}
		return FIRST[prod.noTerminal];
	}
	else if(s.length() == 0)  // 空串
			return set<char>({'@'});
	else if(s.length() == 1){ // 终结符或非终结符
		if(VT.find(s[0]) != VT.end())  // 终结符
			return set<char>({s[0]});
		else { // 非终结符
			if(FIRST[s[0]].size() != 0) return FIRST[s[0]];
			else {
				vector<Prod>::iterator it = find(G.begin(), G.end(), s[0]); // 求出非终结符的产生式
				if(it != G.end()) { // 找到产生式
					set<char> f = first(it->prod);
					FIRST[s[0]].insert(f.begin(), f.end());
				}
				return FIRST[s[0]];
			}
		}
	} else { // 候选式X1X2X3X4...
		set<char> ret;
		for(unsigned int i=0; i<s.length(); ++i) {
			set<char> f = first(string(1, s[i])); // 逐个符号求first(Xi)集
			if(f.find('@') != f.end() && s.length() - 1 != i) { // 发现@
				f.erase(f.find('@')); // 减去@
				ret.insert(f.begin(), f.end()); // 放入first集合
			} else { // 无@，则不需要求下去了
				ret.insert(f.begin(), f.end());
				break;
			}
		}
		return ret;
	}
}

set<char> LL1::follow(const Prod &prod) {
	set<char> folw = FOLLOW[prod.noTerminal]; // 求出产生式X->的follow(X)集合
	if(folw.size() != 0 && !(folw.size() == 1 && folw.find('#') != folw.end())) // 若已求得或为开始符号，直接返回
		return folw;

	// X->aAb
	for(auto p:G) { // 寻找候选式
		for(auto s: p.selection) { // 遍历候选式
			unsigned long loc = 0;
			if((loc = s.find(prod.noTerminal)) != string::npos) { // 找到非终结符
				set<char> f = first(string(s.begin() + loc+1, s.begin() + s.length())); // 求first(b)
				FOLLOW[prod.noTerminal].insert(f.begin(), f.end());
				if(f.find('@') != f.end()) { // 找到@，则follow(X)放入follow(A)
					FOLLOW[prod.noTerminal].erase(FOLLOW[prod.noTerminal].find('@')); // 移除@
					set<char> folw = follow(p); // 递归求follow(X)
					FOLLOW[prod.noTerminal].insert(folw.begin(), folw.end()); // follow(X)加入follow(A)中
				}
			}
		}
	}
	return FOLLOW[prod.noTerminal];
}

void LL1::parseTable() { // X->a
	for(auto prod: G) { // 枚举产生式
		for(auto sel:prod.selection) { // 枚举候选式
			set<char> f = first(sel); // 求first集合
			for(auto terminal: f)
				if(terminal == '@')  // 存在@，将follow(x)的每个终结符放入表中
					for(auto term: FOLLOW[prod.noTerminal])
						M[make_pair(prod.noTerminal, term)] = sel;
				else
					M[make_pair(prod.noTerminal, terminal)] = sel;
		}
	}
}

void LL1::build() {
	for(auto prod: G) first(prod.prod); // 求first集

	FOLLOW[G[0].noTerminal].insert('#'); // 将结束符放入开始符号中
	for(auto prod: G) follow(prod); // 求follow集

	parseTable(); // 预测分析表
	info();
	// 求完表后，将@替换为#
	VT.erase(VT.find('@'));
	VT.insert('#');
	tableInfo();

	return;
}

void LL1::showIndataStack() {
	for(vector<char>::reverse_iterator it = indata.rbegin(); it != indata.rend(); ++it)
		printf("%c", *it);
}

void LL1::showParseStack() {
	for(vector<char>::iterator it = parse.begin(); it != parse.end(); ++it)
		printf("%c", *it);
}

void LL1::loadIndata(const string &s) {
	indata.push_back('#');
	for(int i=s.length()-1; i>=0; --i)
		indata.push_back(s[i]);
}

void LL1::error(int step) {
	printf("{ \"step\": %d, ", step++);

	printf("\"parseStack\": \"ERROR\""
			", \"indataStack\": \"ERROR\""
			", \"production\": \"ERROR\""
			"}\n");
}

void LL1::parser() {
	parse.push_back('#');
	parse.push_back(G[0].noTerminal); // 文法开始符号
	// printf("step\tparseStack\tindataStack\tproduction\n");
	printf("\"Parser\": [");
	int i = 0; // 处理打印的第一个逗号

	int ptop, itop, step = 0;
	string prod = ""; // 候选式
	while((ptop = parse.size() - 1) >= 0) {
		itop = indata.size() - 1;

		printf("%s{ \"step\": %d, ", i++==0?" ":", ", step++);

		printf("\"parseStack\": \"");
		showParseStack();
		printf("\", \"indataStack\": \"");
		showIndataStack();
		printf("\", \"production\": \"%s\"", prod.c_str());
		printf("}\n");

		// begin
		prod = "";
		char X = parse[ptop];
		char curc = indata[itop];
		parse.pop_back();
		if(VT.find(X) != VT.end()) { // 终结符
			if(X != curc) {
				printf("%s", i==0?" ":", ");
				error(step);
				break;
			}
			else {
				indata.pop_back();
				prod = "match " + string(1, X);
			}
		} else if(X == '@')
			continue;
		else { // 终结符
			prod = M[make_pair(X, curc)];
			if(prod.size()) { // 找得到
				if(prod != "@")
					for(int i=prod.size() - 1; i>=0; --i)
						parse.push_back(prod[i]);
			}
			else {
				printf("%s", i==0?" ":", ");
				error(step);
				break;
			}
			prod = string(1, X) + "->" + prod;
		}
	}

	printf("]\n");
	// printf("Parse Success!\n");
}

void LL1::info() {
	printf("\"VT\": [");
	int i = 0; // 处理打印的第一个逗号

	for(auto c: VT) {
		printf("%s\"%c\"", i == 0?" ":", ", c);
		++i;
	}
	printf("], \n");

	i = 0;
	printf("\"VN\": [");
	for(auto c: VN) {
		printf("%s\"%c\"", i == 0?" ":", ", c);
		++i;
	}
	printf("], \n");

	// for(auto p: G)
		// printf("noTerminal: %s\n", p.prod.c_str());

	printf("\"FIRST\": [");
	i = 0;
	for(auto prod:G) {
		printf("\n%c{ \"noTerminal\": \"%c\", \"Terminal\": [", i==0?' ':',', prod.noTerminal);
		int j = 0;
		for(auto c:FIRST[prod.noTerminal]) {
			printf("%s\"%c\"", j == 0?" ":", ", c);
			++j;
		}
		printf("] }");
		++i;
	}
	printf("], \n");

	printf("\"FOLLOW\": [");
	i = 0;
	for(auto prod:G) {
		printf("\n%c{ \"noTerminal\": \"%c\", \"Terminal\": [", i==0?' ':',', prod.noTerminal);
		int j = 0;
		for(auto c:FOLLOW[prod.noTerminal]) {
			printf("%s\"%c\"", j == 0?" ":", ", c);
			++j;
		}
		printf("] }");
		++i;
	}
	printf("], \n");
}

void LL1::tableInfo() {
	printf("\"Table\": {\n");
	printf("\"Header\": [");
	for(auto c:VT)  // 终结符，表头， #号放在最后一列
		if(c != '#')
			printf("\"%c\", ", c);
	printf("\"#\"");
	printf("], \n");

	printf("\"Body\": [");
	int i = 0;
	for(auto prod:G) {
		printf("%s{\"noTerminal\": \"%c\", ", i == 0?" ":", ", prod.noTerminal);
		printf("\"production\": [");
		for(auto c:VT) // 终结符
			if(c != '#')
				printf("\"%s\", ", M[make_pair(prod.noTerminal, c)].c_str());
		printf("\"%s\"", M[make_pair(prod.noTerminal, '#')].c_str());

		printf("]}\n");
		++i;
	}
	printf("] }, \n");
}

void LL1::run() {
	puts("{"); // 输出json，与前端交互
	string in;
	while(cin >> in && in != "#") // 读取文法
		addProd(Prod(in));
	build();
	in = "";
	cin >> in; // 表达式
	loadIndata(in);
	parser();
	puts("}");
}

int main() {
	LL1 ll;
	ll.run();

	return 0;
}
