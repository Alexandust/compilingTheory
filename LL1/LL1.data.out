VT: 
(, ), *, +, -, /, @, i, 
VN: 
E, F, G, S, T, 
noTerminal: E
selections: 
TG
noTerminal: G
selections: 
+TG
-TG
@
noTerminal: T
selections: 
FS
noTerminal: S
selections: 
*FS
/FS
@
noTerminal: F
selections: 
(E)
i
FIRST(E)= {(,i,}
FIRST(G)= {+,-,@,}
FIRST(T)= {(,i,}
FIRST(S)= {*,/,@,}
FIRST(F)= {(,i,}
FOLLOW(E)= {#,),}
FOLLOW(G)= {#,),}
FOLLOW(T)= {#,),+,-,}
FOLLOW(S)= {#,),+,-,}
FOLLOW(F)= {#,),*,+,-,/,}
