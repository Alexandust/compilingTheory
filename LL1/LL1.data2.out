VT: 
#, (, ), *, +, i, 
VN: 
A, B, E, F, T, 
noTerminal: E->TA
noTerminal: A->+TA|@
noTerminal: T->FB
noTerminal: B->*FB|@
noTerminal: F->(E)|i
FIRST(E)= {(,i,}
FIRST(A)= {+,@,}
FIRST(T)= {(,i,}
FIRST(B)= {*,@,}
FIRST(F)= {(,i,}
FOLLOW(E)= {#,),}
FOLLOW(A)= {#,),}
FOLLOW(T)= {#,),+,}
FOLLOW(B)= {#,),+,}
FOLLOW(F)= {#,),*,+,}
Parse Table:
	#	(	)	*	+	i
 E|		TA				TA
 A|	@		@		+TA	
 T|		FB				FB
 B|	@		@	*FB	@	
 F|		(E)				i
i*i+i#
step	parseStack	indataStack	production
0	#E		i*i+i#		
1	#AT		i*i+i#		E->TA
2	#ABF		i*i+i#		T->FB
3	#ABi		i*i+i#		F->i
4	#AB		*i+i#		
5	#ABF*		*i+i#		B->*FB
6	#ABF		i+i#		
7	#ABi		i+i#		F->i
8	#AB		+i#		
9	#A		+i#		B->@
10	#AT+		+i#		A->+TA
11	#AT		i#		
12	#ABF		i#		T->FB
13	#ABi		i#		F->i
14	#AB		#		
15	#A		#		B->@
16	#		#		A->@
Parse Success!
