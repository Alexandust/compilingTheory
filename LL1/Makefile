CXX=clang++
CFLAGS=-O3 --std=c++11
SOURCE = LL1.cpp
DEPS = LL1.h
TARGET=./LL1
TESTIN=LL1.data.in LL1.data2.in

$(TARGET): $(SOURCE) $(DEPS)
	$(CXX) $(CFLAGS) $(SOURCE) -o $(TARGET)
	strip $(TARGET)

clean:
	rm $(TARGET)

test: $(TARGET) $(TESTIN)
	for i in $(TESTIN); \
	do\
		$(TARGET) < $$i > "$$(echo $$i | sed -r 's/(.*)\.in/\1\.out/')";\
	done
