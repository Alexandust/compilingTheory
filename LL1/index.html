<html>
	<head>
		<meta charset="utf-8">
		<title>LL1 parser by netcan</title>
		<script src="../common/jquery.min.js"></script>
		<link href="../common/semantic.min.css" rel="stylesheet">
		<link href="../common/Treant.css" rel="stylesheet">
	</head>
	<body>
		<div class="ui container">
			<h1 class="header">LL1 parser by netcan</h1>
			<div id="parserTree" style="width: 100%; height: 40%"></div>
			<div class="ui grid">
				<!-- 输入串，文法，first集合，follow集合 -->
				<div class="row">
					<div class="four wide column">
						<div class="ui stacked segment">
							<div class="ui form">
								<div class="field">
									<label for="indata">输入串</label>
									<input type="text" id="indata">
								</div>
								<div class="field">
									<label for="grammar">输入文法</label>
									<textarea id="grammar"></textarea>
								</div>
								<button class="small fluid ui submit button" type="submit" id="submit">LL1分析！</button>
							</div>
						</div>
						<div class="ui stacked segment">
							<h4 class="ui horizontal divider header">
								First & Follow
							</h4>
							<div id="ff"></div>
						</div>
					</div>
					<!-- 分析表，分析过程 -->
					<div class="six wide column">
						<div class="ui raised segment">
							<h4 class="ui horizontal divider header">
								Parser
							</h4>
							<table class="ui striped celled table">
								<thead>
									<tr>
										<th>Step</th>
										<th>ParseStack</th>
										<th>InputStack</th>
										<th>Production</th>
									</tr>
								</thead>
								<tbody id="parser">
								</tbody>
							</table>
						</div>
					</div>
					<div class="six wide column">
						<div class="row">
							<div class="ui piled segment">
								<h4 class="ui horizontal divider header">
									Parse Table
								</h4>
								<table class="ui striped celled definition table">
									<thead>
										<tr id="pth">
										</tr>
									</thead>
									<tbody id="ptb">
									</tbody>
								</table>
							</div>
						</div>
					</div>
				</div>
			</div>
		</div>
		<script src="../common/semantic.min.js"></script>
		<script src="../common/raphael.js"></script>
		<script src="../common/Treant.js"></script>
		<script>

			function ParserTree(pt, production) { // 递归构建树
				// console.log(production);
				var p = production.match(/([A-Z])->(.*)/);
				var noTerminal = p[1], Right = p[2];
				if(typeof pt['text'] === "undefined") {
					pt['text'] = {
						name: noTerminal
					}
					pt['children'] = [];
					for(var i=0; i< Right.length; ++i)
						pt['children'].push({text: { name: Right[i] }});
				} else {
					if(pt['text']['name'] != noTerminal) {
						if(typeof pt['children'] != "undefined")
							for(var i=0, len=pt['children'].length;  i < len; ++i) {
								var child = pt['children'][i];
								ParserTree(child, production);
							}
					}
					else {
						if(typeof pt['children'] === "undefined") {
							pt['children'] = [];
							for(var i=0; i< Right.length; ++i) {
								pt['children'].push({text: { name: Right[i] }});
							}
						} else {
							for(var i=0, len=pt['children'].length;  i < len; ++i) {
								var child = pt['children'][i];
								if(child['text']['name'] == noTerminal)
									ParserTree(child, production);
							}

						}
					}
				}
			}

			$("#submit").click(function() {
				var data = $('#grammar').val() + '\n#\n' + $('#indata').val();
				//	console.log(data);
				$.ajax({
					url: "LL1.php",
					datatype: "json",
					method: "POST",
					data: {"data": data},
					success: function(res, status, jqxhr) {
						console.log(res);
						// first and follow
						var ff = '';
						for(var i = 0, len = res['FIRST'].length; i<len; ++i) {
							var first = res['FIRST'][i];
							ff += '<p>FIRST('+ first['noTerminal'] +') = {';
								for(var j = 0, jlen = first['Terminal'].length; j < jlen; ++j)
									ff += (j==0?" ":", ") + first['Terminal'][j];
							ff += ' }</p>'
						}
						for(var i = 0, len = res['FOLLOW'].length; i<len; ++i) {
							var follow = res['FIRST'][i];
							ff += '<p>FOLLOW('+ follow['noTerminal'] +') = {';
								for(var j = 0, jlen = follow['Terminal'].length; j < jlen; ++j)
									ff += (j==0?" ":", ") + follow['Terminal'][j];
							ff += ' }</p>'
						}
						$('#ff').html(ff);

						// parse table
						var pth = '<th></th>';
						var table = res['Table'];
						for(var i=0, len = table['Header'].length; i < len; ++i)
							pth += '<th>'+ table['Header'][i] +'</th>';
						var ptb = '';
						for(var i=0, len = table['Body'].length; i < len; ++i) {
							var row = table['Body'][i];
							ptb += '<tr><td>' + row['noTerminal'] + '</td>';
							for(var j=0, jlen = row['production'].length; j < jlen; ++j)
								ptb += '<td>' + row['production'][j] + '</td>';
							ptb += '</tr>'
						}

						// console.log(ptb);
						$('#pth').html(pth);
						$('#ptb').html(ptb);

						// parser
						var parser = '';
						var parserTree = {};
						parserTree['chart'] = {
							container: "#parserTree",
						};
						parserTree['nodeStructure'] = { };

						var isMatch = new RegExp('match');
						for(var i=0, len = res['Parser'].length; i < len; ++i) {
							var row = res['Parser'][i];
							var production = res['Parser'][(i+1) % len]['production'];
							if(! isMatch.test(production) && production.length != 0 && production != "ERROR")
								ParserTree(parserTree['nodeStructure'], production);

							parser += '<tr class="' + (row['parseStack'] == "ERROR"?"negative":"") + '"><td>' + i + '</td><td>' + row['parseStack'] + '</td>' +
								'<td>' + row['indataStack'] + '</td>' +
								'<td>' + production + '</td>' +
								'</th></tr>';
						}
						// console.log(parserTree);
						$('#parser').html(parser);

						$('#parserTree').html("");
						Treant(parserTree, function() {
							$('#parserTree').css('height', $('svg').height())
						});
					}
				})
			});

			$(document).ready(function() {
				$('#indata').val('i*i+i*(i+i)');
				$('#grammar').val(
						"E->TG\n" +
						"G->+TG|-TG\n" +
						"G->@\n" +
						"T->FS\n" +
						"S->*FS|/FS\n" +
						"S->@\n" +
						"F->(E)\n" +
						"F->i"
						);
				$('#submit').click();
			});
		</script>

	</body>
</html>
