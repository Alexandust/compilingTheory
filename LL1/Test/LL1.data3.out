{
"VT": [ "@", "a", "b", "d"], 
"VN": [ "A", "S"], 
"FIRST": [
 { "noTerminal": "S", "Terminal": [ "a", "d"] }
,{ "noTerminal": "A", "Terminal": [ "@", "b"] }], 
"FOLLOW": [
 { "noTerminal": "S", "Terminal": [ "#", "a", "d"] }
,{ "noTerminal": "A", "Terminal": [ "#", "a", "d"] }], 
"Table": {
"Header": ["a", "b", "d", "#"], 
"Body": [ {"noTerminal": "S", "production": ["aA", "", "d", ""]}
, {"noTerminal": "A", "production": ["@", "bAS", "@", "@"]}
] }, 
"Parser": [ { "step": 0, "parseStack": "#S", "indataStack": "#", "production": ""}
, { "step": 1, "parseStack": "ERROR", "indataStack": "ERROR", "production": "ERROR"}
]
}
